### ISSUE TITLE
AnimatedImageView crash

### ISSUE THREAD

**Issue by MalcolmnDEV** [onevcat/Kingfisher#1844]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [x] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [x] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [x] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

#### What

Since the update to Kingfisher 7.0.0 users have been reporting several crashes inside our app. Our crash reporting software captured the stack traces and it seems to be inside the Kingfisher iOS SDK.

- AnimatedImageView.Animator.loadFrame(at:)
Stack Trace:

Crashed: com.onevcat.Kingfisher.Animator.preloadQueue
0  libsystem_platform.dylib       0x1a44 _platform_memmove + 548
1  CoreFoundation                 0xb32f8 CFDataGetBytes + 328
2  ImageIO                        0x1500 CGImageGetImageSource + 152
3  UIKitCore                      0x203024 -[_UIImageCGImageContent initWithCGImage:scale:] + 36
4  UIKitCore                      0x18c3ec -[UIImage initWithCGImage:scale:orientation:] + 72
5  Woebot                         0x266520 AnimatedImageView.Animator.loadFrame(at:) + 4336739616 (<compiler-generated>:4336739616)
6  Woebot                         0x2667b8 AnimatedImageView.Animator.updatePreloadedFrames() + 575 (AnimatedImageView.swift:575)
7  Woebot                         0x1d51f0 thunk for @escaping @callee_guaranteed () -> () + 4336144880 (<compiler-generated>:4336144880)
8  libdispatch.dylib              0x632ec _dispatch_call_block_and_release + 24
9  libdispatch.dylib              0x642f0 _dispatch_client_callout + 16
10 libdispatch.dylib              0x3f410 _dispatch_lane_serial_drain$VARIANT$armv81 + 596
11 libdispatch.dylib              0x3feb8 _dispatch_lane_invoke$VARIANT$armv81 + 388
12 libdispatch.dylib              0x4976c _dispatch_workloop_worker_thread + 616
13 libsystem_pthread.dylib        0xf38 _pthread_wqthread + 284
14 libsystem_pthread.dylib        0xaa4 start_wqthread + 8

- AnimatedImageView.Animator.updatePreloadedFrames()
Stack Trace:

Crashed: com.onevcat.Kingfisher.Animator.preloadQueue
0  CoreFoundation                 0xbc124 CFDataGetBytes + 156
1  ImageIO                        0x2700 CGImageGetImageSource + 156
2  UIKitCore                      0x215c74 -[_UIImageCGImageContent initWithCGImage:scale:] + 40
3  UIKitCore                      0x19cee0 -[UIImage initWithCGImage:scale:orientation:] + 76
4  Woebot                         0x266520 (Missing)
5  Woebot                         0x2667b8 AnimatedImageView.Animator.updatePreloadedFrames() + 575 (AnimatedImageView.swift:575)
6  Woebot                         0x1d51f0 thunk for @escaping @callee_guaranteed () -> () + 4309012976 (<compiler-generated>:4309012976)
7  libdispatch.dylib              0x1c04 _dispatch_call_block_and_release + 32
8  libdispatch.dylib              0x3950 _dispatch_client_callout + 20
9  libdispatch.dylib              0xb0ac _dispatch_lane_serial_drain + 664
10 libdispatch.dylib              0xbc10 _dispatch_lane_invoke + 392
11 libdispatch.dylib              0x16318 _dispatch_workloop_worker_thread + 656
12 libsystem_pthread.dylib        0x11b0 _pthread_wqthread + 288
13 libsystem_pthread.dylib        0xf50 start_wqthread + 8

AnimatedImageView is used inside the app to display an image full screen.

#### Reproduce

```swift
class FullScreenImageViewController: UIViewController {
    private var imageView: AnimatedImageView!
    private var scrollView: UIScrollView!

    private static let animationDuration: TimeInterval = 0.15

    func showFullScreen(animatedImageView: AnimatedImageView) {
        self.setupScrollView()

        self.imageView = AnimatedImageView(image: animatedImageView.image)
        self.imageView.contentMode = .scaleAspectFit

        guard let delegate = UIApplication.shared.delegate,
            let wrappedWindow = delegate.window,
            let window = wrappedWindow,
            let rootViewController = window.rootViewController,
            let imageSize = animatedImageView.image?.size
        else { return }

        self.scrollView.addSubview(self.imageView)
        rootViewController.view.addSubview(self.scrollView)

        self.imageView.translatesAutoresizingMaskIntoConstraints = false
        self.imageView.widthAnchor.constraint(lessThanOrEqualTo: self.scrollView.widthAnchor).isActive = true
        self.imageView.heightAnchor.constraint(equalTo: self.imageView.widthAnchor, multiplier: imageSize.height / imageSize.width).isActive = true
        self.imageView.heightAnchor.constraint(lessThanOrEqualTo: self.scrollView.heightAnchor).isActive = true
        self.imageView.centerXAnchor.constraint(equalTo: self.scrollView.centerXAnchor).isActive = true
        self.imageView.centerYAnchor.constraint(equalTo: self.scrollView.centerYAnchor).isActive = true

        UIView.animate(withDuration: FullScreenImageViewController.animationDuration) {
            self.scrollView.alpha = 1
        }
    }

    private func setupScrollView() {
        self.scrollView = UIScrollView(frame: UIScreen.main.bounds)
        self.scrollView.maximumZoomScale = 2.0
        self.scrollView.delegate = self
        self.scrollView.backgroundColor = .black
        self.scrollView.alpha = 0
        self.scrollView.showsVerticalScrollIndicator = false
        self.scrollView.showsHorizontalScrollIndicator = false

        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(self.dismissFullscreenImage))
        self.scrollView.addGestureRecognizer(tapGestureRecognizer)
        let swipeGestureRecognizer = UISwipeGestureRecognizer(target: self, action: #selector(self.dismissFullscreenImage))
        swipeGestureRecognizer.direction = .down
        self.scrollView.addGestureRecognizer(swipeGestureRecognizer)
        let doubleTapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(self.didDoubleTap))
        doubleTapGestureRecognizer.numberOfTapsRequired = 2
        self.scrollView.addGestureRecognizer(doubleTapGestureRecognizer)
        tapGestureRecognizer.require(toFail: doubleTapGestureRecognizer)
    }

    @objc private func dismissFullscreenImage(_ sender: UIGestureRecognizer) {
        UIView.animate(withDuration: FullScreenImageViewController.animationDuration, animations: {
            sender.view?.alpha = 0
        }, completion: { _ in
            sender.view?.removeFromSuperview()
            self.removeFromParent()
        })
    }

    @objc private func didDoubleTap(_ sender: UITapGestureRecognizer) {
        guard self.scrollView.zoomScale == 1 else {
            self.scrollView.setZoomScale(1, animated: true)
            return
        }
        let zoomRect = self.zoomRect(forScale: self.scrollView.maximumZoomScale, center: sender.location(in: sender.view))
        self.scrollView.zoom(to: zoomRect, animated: true)
    }

    private func zoomRect(forScale scale: CGFloat, center: CGPoint) -> CGRect {
        var zoomRect: CGRect = .zero
        zoomRect.size.height = imageView.frame.size.height / scale
        zoomRect.size.width = imageView.frame.size.width / scale
        let newCenter = self.scrollView.convert(center, from: imageView)
        zoomRect.origin.x = newCenter.x - (zoomRect.size.width / 2.0)
        zoomRect.origin.y = newCenter.y - (zoomRect.size.height / 2.0)
        return zoomRect
    }
}
```

**Comment by tkotf61412** [comment]
same problem. AnimatedImageView crash in iOS15

**Comment by yayang** [comment]
> I'm experiencing this occasional crash on iOS 15 too using both device and simulator running Kingfisher 7.0 (seems like iOS 15 is the culprit).
> I find that setting `framePreloadCount = .max` "fixes" the crash, but at a cost of more memory consumption

works for me by now, thanks

**Comment by MalcolmnDEV** [comment]
The crash is still happening on my end enough though I added `framePreloadCount = .max`.

I have also not been able to trigger the crash myself but our crash reports are showing several crashes every day.

Crashed: com.onevcat.Kingfisher.Animator.preloadQueue
0  libsystem_platform.dylib       0x1a44 _platform_memmove + 548
1  CoreFoundation                 0xb32e8 CFDataGetBytes + 328
2  ImageIO                        0x1c64 CGImageGetImageSource + 152
3  UIKitCore                      0x202bec -[_UIImageCGImageContent initWithCGImage:scale:] + 36
4  UIKitCore                      0x18c718 -[UIImage initWithCGImage:scale:orientation:] + 72
5  Woebot                         0x232bc8 AnimatedImageView.Animator.loadFrame(at:) + 4341296072 (<compiler-generated>:4341296072)
6  Woebot                         0x232e60 AnimatedImageView.Animator.updatePreloadedFrames() + 579 (AnimatedImageView.swift:579)
7  Woebot                         0x19d1c4 thunk for @escaping @callee_guaranteed () -> () + 4340683204 (<compiler-generated>:4340683204)
8  libdispatch.dylib              0x63194 _dispatch_call_block_and_release + 24
9  libdispatch.dylib              0x64198 _dispatch_client_callout + 16
10 libdispatch.dylib              0x3f13c _dispatch_lane_serial_drain$VARIANT$armv81 + 604
11 libdispatch.dylib              0x3fbf4 _dispatch_lane_invoke$VARIANT$armv81 + 388
12 libdispatch.dylib              0x495cc _dispatch_workloop_worker_thread + 616
13 libsystem_pthread.dylib        0x4e84 _pthread_wqthread + 284
14 libsystem_pthread.dylib        0x49f0 start_wqthread + 8

@onevcat  I am not sure what other information I can give you other than the crash report and the code I added in the above description

**Comment by onevcat** [comment]
@MalcolmnDEV Thanks for the information. May I know if you have any information about the displayed GIF file? Is it a user generated one or an image you created and are having?

This crash happens inside `CGImageGetImageSource` and data bytes getting. Kingfisher did not even touch the `AnimatedImageView` recently so I suspect it is some change in iOS 15's ImageIO..But at least we need a way to reproduce before we can say what exactly can be wrong.

**Comment by MalcolmnDEV** [comment]
@onevcat The GIF file comes from the backend. Here are a few that were used in the last month.
[gifs.zip](https://github.com/onevcat/Kingfisher/files/7555836/gifs.zip)

**Comment by zhao560** [comment]
![image](https://user-images.githubusercontent.com/26494894/142978469-983b52b9-7bc6-4f3e-bb76-8f03f49beefe.png)

**Comment by Soksijada** [comment]
Any news for this?

**Comment by MalcolmnDEV** [comment]
@onevcat Any news on this it happened 143 times over the holidays on our App

**Comment by onevcat** [comment]
Ummm. Maybe related to this https://github.com/SDWebImage/SDWebImage/pull/3308

I will check how and whether it also affects Kingfisher.

**Comment by onevcat** [comment]
Hmmm...Tried a bit but seems to the same with SDWebImage's crash. Unlike in SDWebImage, Kingfisher does not holds the decoder even when multiple image views are showing the same image.

@MalcolmnDEV And from your original description, it seems that this crash happens when

> AnimatedImageView is used inside the app to display an image full screen.

so can I say that in your situation, only one `AnimatedImageView` will be existing at the same time?

**Comment by MalcolmnDEV** [comment]
@onevcat Sorry about the limit crash report, that's all the info crashlytics provides. I have not been able to replicate it on my end.

I can say that its on iOS 15 and iPad OS 15.

There will be 2 `AnimatedImageView` when it comes to full screen. The full screen version gets display ontop of the tableview that holds the original in a TableViewCell .

Note that its Gifs not static images

**Comment by MalcolmnDEV** [comment]
@onevcat this is the stack trace where it looks like it does crash with `_UIImageCGImageContent `

`Crashed: com.onevcat.Kingfisher.Animator.preloadQueue
0  libobjc.A.dylib                0x1014 objc_msgSend + 20
1  ImageIO                        0x1c64 CGImageGetImageSource + 152
2  UIKitCore                      0x202bec -[_UIImageCGImageContent initWithCGImage:scale:] + 36
3  UIKitCore                      0x18c718 -[UIImage initWithCGImage:scale:orientation:] + 72`

**Comment by zhongxunchao** [comment]
Perhaps not the same crash: 

Thread 11 name:   Dispatch queue: com.onevcat.Kingfisher.Animator.preloadQueue
Thread 11 Crashed:
0   CoreFoundation                	       0x181364114 CFDataGetBytes + 156
1   ImageIO                       	       0x182dd1d48 CGImageGetImageSource + 156
2   ImageIO                       	       0x182dd1d48 CGImageGetImageSource + 156
3   UIKitCore                     	       0x18391e954 -[_UIImageCGImageContent dealloc] + 48
4   UIKitCore                     	       0x183997484 -[UIImage .cxx_destruct] + 32
5   libobjc.A.dylib               	       0x1999a7f18 object_cxxDestructFromClass(objc_object*, objc_class*) + 116
6   libobjc.A.dylib               	       0x1999a4c14 objc_destructInstance + 80
7   libobjc.A.dylib               	       0x1999ae338 _objc_rootDealloc + 80
8   UIKitCore                     	       0x1838fd4ac -[UIImage dealloc] + 76
9   Kingfisher                    	       0x107743378 specialized SafeArray.subscript.setter + 184
10  Kingfisher                    	       0x107742e14 AnimatedImageView.Animator.updatePreloadedFrames() + 208
11  Kingfisher                    	       0x107747840 thunk for @escaping @callee_guaranteed () -> () + 20
12  libdispatch.dylib             	       0x180fb1914 _dispatch_call_block_and_release + 32
13  libdispatch.dylib             	       0x180fb3660 _dispatch_client_callout + 20
14  libdispatch.dylib             	       0x180fbade4 _dispatch_lane_serial_drain + 672
15  libdispatch.dylib             	       0x180fbb958 _dispatch_lane_invoke + 392
16  libdispatch.dylib             	       0x180fc61a8 _dispatch_workloop_worker_thread + 656
17  libsystem_pthread.dylib       	       0x1f1eca0f4 _pthread_wqthread + 288
18  libsystem_pthread.dylib       	       0x1f1ec9e94 start_wqthread + 8

**Comment by onevcat** [comment]
Ummm...

The crash above is not the same, but I guess similar. It is in `-[_UIImageCGImageContent dealloc]` and again, related to `_UIImageCGImageContent`. It would be a great loss if the `_UIImageCGImageContent` related method are not able to run on a background thread, or the get bytes is not thread safe.

**Comment by peter-Kings** [comment]
最上面的crash遇到了，还有一个相似的SafeArray崩溃，初步定位prepareFramesAsynchronously方法内。
以下是友盟的解析结果，取到本地解析也好不了多少，勉强看看？

Thread 0 Crashed:
0   CoreFoundation                      0x00000001804d4114 CFDataGetBytes
1   ImageIO                             0x0000000181f41d48 CGImageGetImageSource
2   ImageIO                             0x0000000181f41d48 CGImageGetImageSource
3   UIKitCore                           0x0000000182a8e954 -[_UIImageCGImageContent dealloc]
4   UIKitCore                           0x0000000182b07484 -[UIImage .cxx_destruct]
5   libobjc.A.dylib                     0x0000000198b17f18 object_cxxDestructFromClass(objc_object*, objc_class*)
6   libobjc.A.dylib                     0x0000000198b14c14 objc_destructInstance
7   libobjc.A.dylib                     0x0000000198b1e338 _objc_rootDealloc
8   UIKitCore                           0x0000000182a6d4ac -[UIImage dealloc]
9   libswiftCore.dylib                  0x00000001851cb9c0 swift_arrayDestroy
10  libswiftCore.dylib                  0x0000000184ee85d8 $ss23_ContiguousArrayStorageCfD
11  libswiftCore.dylib                  0x00000001851d8de0 _swift_release_dealloc
12  mama                                0x00000001051072e8 $s10Kingfisher9SafeArrayCfD + [<compiler-generated> : 0] 
13  libswiftCore.dylib                  0x00000001851d8de0 _swift_release_dealloc
14  mama                                0x0000000105105f6c $s10Kingfisher17AnimatedImageViewC8AnimatorCfD + [<compiler-generated> : 0] 
15  libswiftCore.dylib                  0x00000001851d8de0 _swift_release_dealloc
16  libswiftCore.dylib                  0x00000001851d9d84 bool swift::HeapObjectSideTableEntry::decrementStrong<(swift::PerformDeinit)1>(unsigned int)
17  mama                                0x00000001051048c0 $s10Kingfisher17AnimatedImageViewCfETo + [AnimatedImageView.swift : 0] 
18  libobjc.A.dylib                     0x0000000198b17f18 object_cxxDestructFromClass(objc_object*, objc_class*)
19  libobjc.A.dylib                     0x0000000198b14c14 objc_destructInstance
20  libobjc.A.dylib                     0x0000000198b1e338 _objc_rootDealloc
21  UIKitCore                           0x0000000182bbeea0 -[UIResponder dealloc]
22  UIKitCore                           0x0000000182ab660c -[UIView dealloc]
23  mama                                0x0000000105431c90 __66-[RXObjCRuntime(InfrastructureMethods) swizzleDeallocating:error:]_block_invoke_2.111 + [_RXObjCRuntime.m : 508] 
24  UIKitCore                           0x0000000182af748c -[UIImageView dealloc]
25  mama                                0x0000000105104870 $s10Kingfisher17AnimatedImageViewCfDTo + [<compiler-generated> : 0] 
26  libobjc.A.dylib                     0x0000000198b1755c AutoreleasePoolPage::releaseUntil(objc_object**)
27  libobjc.A.dylib                     0x0000000198b13928 objc_autoreleasePoolPop
28  CoreFoundation                      0x000000018042c59c _CFAutoreleasePoolPop
29  CoreFoundation                      0x0000000180425298 __CFRunLoopPerCalloutARPEnd
30  CoreFoundation                      0x0000000180423ef0 __CFRunLoopRun
31  CoreFoundation                      0x00000001804373b8 CFRunLoopRunSpecific
32  GraphicsServices                    0x000000019bdc738c GSEventRunModal
33  UIKitCore                           0x0000000182dd76a8 -[UIApplication _run]
34  UIKitCore                           0x0000000182b567f4 UIApplicationMain
35  mama                                0x00000001040f3f20 main + [AppDelegate.swift : 30] 

Thread 24:
0   libswiftCore.dylib                  0x00000001851eecb4 swift::MetadataCacheKey::operator==(swift::MetadataCacheKey) const
1   libswiftCore.dylib                  0x00000001851e0ff0 _swift_getGenericMetadata(swift::MetadataRequest, void const* const*, swift::TargetTypeContextDescriptor<swift::InProcess> const*)
2   libswiftCore.dylib                  0x00000001851e0ff0 _swift_getGenericMetadata(swift::MetadataRequest, void const* const*, swift::TargetTypeContextDescriptor<swift::InProcess> const*)
3   libswiftCore.dylib                  0x00000001851c1540 __swift_instantiateGenericMetadata
4   libswiftFoundation.dylib            0x0000000184b013f8 $sSD10FoundationE34_conditionallyBridgeFromObjectiveC_6resultSbSo12NSDictionaryC_SDyxq_GSgztFZSS_ypTg5
5   libswiftCore.dylib                  0x00000001851d3074 _tryCastFromClassToObjCBridgeable(swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, void*, swift::TargetMetadata<swift::InProcess> const*&, swift::TargetMetadata<swift::InProcess> const*&, bool, bool, _ObjectiveCBridgeableWitnessTable const*, swift::TargetMetadata<swift::InProcess> const*)
6   libswiftCore.dylib                  0x00000001851d1828 tryCast(swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::TargetMetadata<swift::InProcess> const*&, swift::TargetMetadata<swift::InProcess> const*&, bool, bool)
7   libswiftCore.dylib                  0x00000001851d1454 tryCast(swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::TargetMetadata<swift::InProcess> const*&, swift::TargetMetadata<swift::InProcess> const*&, bool, bool)
8   libswiftCore.dylib                  0x00000001851d0d2c swift_dynamicCastImpl(swift::OpaqueValue*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*, swift::TargetMetadata<swift::InProcess> const*, swift::DynamicCastFlags)
9   mama                                0x0000000105106574 $s10Kingfisher17AnimatedImageViewC8AnimatorC05setupB6Frames33_F66BEC896F1F38CAC29BEFA58C261558LLyyF + [GIFAnimatedImage.swift : 118] 
10  mama                                0x0000000105108aec $s10Kingfisher17AnimatedImageViewC8AnimatorC27prepareFramesAsynchronouslyyyFyycfU_TA + [<compiler-generated> : 0] 
11  mama                                0x00000001041f2314 $sIeg_IeyB_TR + [<compiler-generated> : 0] 
12  libdispatch.dylib                   0x0000000180121914 _dispatch_call_block_and_release
13  libdispatch.dylib                   0x0000000180123660 _dispatch_client_callout
14  libdispatch.dylib                   0x000000018012ade4 _dispatch_lane_serial_drain
15  libdispatch.dylib                   0x000000018012b958 _dispatch_lane_invoke
16  libdispatch.dylib                   0x00000001801361a8 _dispatch_workloop_worker_thread
17  libsystem_pthread.dylib             0x00000001f103a0f4 _pthread_wqthread

**Comment by rafalszastok** [comment]
> I'm experiencing this occasional crash on iOS 15 too using both device and simulator running Kingfisher 7.0 (seems like iOS 15 is the culprit). I find that setting `framePreloadCount = .max` "fixes" the crash, but at a cost of more memory consumption

I can confirm that 'equalizer.framePreloadCount = .max' fixes the problem in my case. My code to reproduce the crash below. Note: you need to add to project gif image called `levels_light@3x.gif`

```
final class PresentingViewController: UIViewController {

    var equalizerArray: [AnimatedImageView] = []

    override func viewDidLoad() {
        super.viewDidLoad()

        for x in [50, 100, 150, 200, 250] {
            for y in [-50, -100, -150, -200, -250, -300, -350] {
                let frame = CGRect(x: x, y: y, width: 30, height: 30)
                let equalizer = AnimatedImageView()
                equalizer.translatesAutoresizingMaskIntoConstraints = false
                setup(equalizer: equalizer, frame: frame)
                equalizerArray.append(equalizer)
            }
        }

    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        for equalizer in equalizerArray {
            equalizer.startAnimating()
        }
    }

    private func setup(equalizer: AnimatedImageView, frame: CGRect) {

        equalizer.needsPrescaling = false

        if let path = Bundle.main.path(
            forResource: "levels_light@3x",
            ofType: "gif") {
            equalizer.kf.setImage(with: URL(fileURLWithPath: path))
        }
        view.addSubview(equalizer)

        let x = frame.origin.x
        let y = frame.origin.y
        let height = frame.size.height
        let width = frame.size.width

        NSLayoutConstraint
            .activate([
                equalizer.heightAnchor.constraint(equalToConstant: height),
                equalizer.widthAnchor.constraint(equalToConstant: width),
                equalizer.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: x),
                equalizer.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: y)
            ])

    }
}
```

**Comment by ufosky** [comment]
这里提到的2个crash栈我都遇到了，这是其中一个截图
<img width="1290" alt="image" src="https://user-images.githubusercontent.com/801838/174423040-df457ef9-2286-4aa6-b8c2-cc0a96188699.png">

**Comment by peter-Kings** [comment]
你的邮件已收到，我会尽快处理并给予回复！！！

**Comment by ufosky** [comment]
> 这里提到的2个crash栈我都遇到了，这是其中一个截图 <img alt="image" width="1290" src="https://user-images.githubusercontent.com/801838/174423040-df457ef9-2286-4aa6-b8c2-cc0a96188699.png">

@onevcat  提交了一个PR修复crash，我截图的crash比较容易出现，展示几个gif图片后放那不动就会出现，修复后没出现过了。下面这个crash我也出现过，但是比较难出现，修复后的代码运行也没出现过了，不确定是否修复
> ![image](https://user-images.githubusercontent.com/15356469/137257251-a65b7a17-fffa-4e3b-b283-26315376ac56.png)

**Comment by ufosky** [comment]
This crash still occurred
<img width="1446" alt="image" src="https://user-images.githubusercontent.com/801838/175181192-1e53833b-fbbf-4f3e-8de4-e3d738b66ca3.png">

**Comment by onevcat** [comment]
That PR was merged, but it was not yet released as a tag. So please make sure you are pointing to either your fix branch or the master branch. 

(But I have to say it is still possible to have this crash. It seems `_UIImageCGImageContent` does not read data in a thread safe way, but since it is not documented anywhere and it might introduce a huge impact to move to a global serial queue, I am now hesitating on what to do with it...:( )

**Comment by nitikorndev** [comment]
For me I switch to using UIImageView way  and its helped

/// AnimatedImageView represents a subclass of `UIImageView` for displaying animated image.
/// Different from showing animated image in a normal `UIImageView` (which load all frames at one time),
/// `AnimatedImageView` only tries to load several frames (defined by `framePreloadCount`) to reduce memory usage.
/// It provides a tradeoff between memory usage and CPU time. If you have a memory issue when using a normal image

**Comment by ufosky** [comment]
I think the construction and destruction of the UIImage from cgimage should both be in the main thread.

**Comment by dreampiggy** [comment]
This is the same issue as https://github.com/SDWebImage/SDWebImage/issues/3273#issuecomment-1203696655

From iOS 15+, `CGImage` keep a strong reference to the `CGImageSource` wihch produce it.
So, if you create CGImage in one thread, and do not release that CGImageSource, once UIKit.UIImageView render the CGImage on the main queue, this may cause thread safe conflict.

One possible solution is to call private API `CGImageSetImageSource` to strip out, or have to workaround this by using CoreGraphics to redraw and copy the CGImage.

**Comment by onevcat** [comment]
@dreampiggy It is really useful and thank you for sharing this to me! I will see if it is possible to handle it in the same way in Kingfisher too.

**Comment by quyentrinh** [comment]
I also got this problem. Is it fixed in the latest version?

**Comment by peter-Kings** [comment]
你的邮件已收到，我会尽快处理并给予回复！！！

**Comment by Creolophus** [comment]
Hi, any update?

**Comment by a7340103** [comment]
Is is fixed in version 6.3.1? To some reason, my project need support iOS10+

**Comment by lehom** [comment]
![image](https://user-images.githubusercontent.com/10010155/212447529-37a2af43-4a23-4508-8cc5-2e63071c6602.png)
The crash still reproduce in version 7.4.1

**Comment by onevcat** [comment]
@lehom 

This seems to be another issue the `cgImage` was not copied/decoded correctly, so it falls back to the same implementation as before the patch was applied.

Any idea or information about the image itself?

**Comment by lehom** [comment]
> 
### Reproduce the crash

I can reproduce this crash by multiple GIFs accessing the same url, change the code in GIFViewController.swift on Kingfisher-Demo Project  as follows:

```swift
class GIFViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var animatedImageView: AnimatedImageView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
//        let url = ImageLoader.gifImageURLs.last!
        
//        // Should need to use different cache key to prevent data overwritten by each other.
//        KF.url(url, cacheKey: "\(url)-imageview").set(to: imageView)
//        KF.url(url, cacheKey: "\(url)-animated_imageview").set(to: animatedImageView)
        
        let url = URL(string: "https://raw.githubusercontent.com/onevcat/Kingfisher-TestImages/master/DemoAppImage/GIF/GifHeavy.gif")
        
        (0...25).forEach { idx in
            
            let imageView = AnimatedImageView()
            imageView.frame = CGRect(x: 0, y: 10 + idx*100, width: 100, height: 100)
            
            let imageView1 = AnimatedImageView()
            imageView1.frame = CGRect(x: 110, y: 10 + idx*100, width: 100, height: 100)
            
            
            let imageView2 = AnimatedImageView()
            imageView2.frame = CGRect(x: 220, y: 10 + idx*100, width: 100, height: 100)
            
            
            view.addSubview(imageView)
            view.addSubview(imageView1)
            view.addSubview(imageView2)

            KF.url(url, cacheKey: "\(url)-animated_imageview\(UUID().uuidString)").set(to: imageView)
            KF.url(url, cacheKey: "\(url)-animated_imageview\(UUID().uuidString)").set(to: imageView1)
            KF.url(url, cacheKey: "\(url)-animated_imageview\(UUID().uuidString)").set(to: imageView2)

        }
    }
}
```
### Fix
1. I guess You tried to solve this crash before by the way as SDWebImage [#3387MR](https://github.com/SDWebImage/SDWebImage/pull/3387). But I found a problem with your code

![image](https://user-images.githubusercontent.com/10010155/212789902-c8359fd5-692f-4d84-b099-51d34e34e2c3.png)

the `GraphicsContext.current(size: imageSize, scale: imageScale, inverting: true, cgImage: cgImage)` return nil. I think because `GraphicsContext.begin(size: imageSize, scale: imageScale)` and `GraphicsContext.current(size: imageSize, scale: imageScale, inverting: true, cgImage: cgImage)` not call on the same thread, result in `UIGraphicsGetCurrentContext() = nil`.

2. We fixed by this way [#3245](https://github.com/SDWebImage/SDWebImage/pull/3425)  to strip down the CGImageSourceRef from CGImageRef, change the code on AnimatedImageView.swift in Kingfisher Project:

```swift
           .........
            if #available(iOS 15, tvOS 15, *) {
                // From iOS 15, a plain image loading causes iOS calling `-[_UIImageCGImageContent initWithCGImage:scale:]`
                // in ImageIO, which holds the image ref on the creating thread.
                // To get a workaround, create another image ref and use that to create the final image. This leads to
                // some performance loss, but there is little we can do.
                // https://github.com/onevcat/Kingfisher/issues/1844
//                guard let context = GraphicsContext.current(size: imageSize, scale: imageScale, inverting: true, cgImage: cgImage),
//                      let decodedImageRef = cgImage.decoded(on: context, scale: imageScale)
//                else {
//                    return KFCrossPlatformImage(cgImage: cgImage)
//                }
                guard let newImageRef = Animator.SDCGImageCreateCopy(cgImage: cgImage) else {
                    return KFCrossPlatformImage(cgImage: cgImage)
                }
                return KFCrossPlatformImage(cgImage: newImageRef)
            }
        ...........

        static func SDCGImageCreateCopy(cgImage: CGImage?) -> CGImage? {
            guard let cgImage = cgImage,
                  let space = cgImage.colorSpace,
                  let provider = cgImage.dataProvider
            else {
                return nil
            }
            
            let width = cgImage.width
            let height = cgImage.height
            let bitsPerComponent = cgImage.bitsPerComponent
            let bitsPerPixel = cgImage.bitsPerPixel
            let bytesPerRow = cgImage.bytesPerRow
            let bitmapInfo = cgImage.bitmapInfo
            let decode = cgImage.decode
            let shouldInterpolate = cgImage.shouldInterpolate
            let intent = cgImage.renderingIntent
            let newImage = CGImage(width: width, height: height, bitsPerComponent: bitsPerComponent, bitsPerPixel: bitsPerPixel, bytesPerRow: bytesPerRow, space: space, bitmapInfo: bitmapInfo, provider: provider, decode: decode, shouldInterpolate: shouldInterpolate, intent: intent)
            return newImage;
        }
```
But I think you should have a better way, consider memory and cpu. Thanks!

### Refer to

[探秘越来越复杂的ImageIO框架](https://dreampiggy.com/2022/11/07/%E6%8E%A2%E7%A7%98%E8%B6%8A%E6%9D%A5%E8%B6%8A%E5%A4%8D%E6%9D%82%E7%9A%84ImageIO%E6%A1%86%E6%9E%B6/)

**Comment by desolateug** [comment]
@onevcat The same problem as @lehom.

**Comment by eyuxin** [comment]
Any progress?

**Comment by onevcat** [comment]
I found the actual issue and applied a fix in the latest [version 7.10.2](https://github.com/onevcat/Kingfisher/releases/tag/7.10.2). Please try to upgrade to get the crash gone.

The fix itself is not yet ideal and brings some performance regression. But at least it won't crash anymore. I will see if there is a better way later.
